Improve hypersimp to give results for more hypergeometric terms

Harikrishna-Srinivasan opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #27975 

<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

#### Brief description of what is fixed or changed
`hypersimp` gives `None` for many hypergeomtric terms, which is fixed in this PR.

#### Other comments

For example, `(2*k + 1)*binomial(n, k)` returns `None` instead of
```python
-(k - n)*(2*k + 3)/((k + 1)*(2*k + 1))
```

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* simplify
  * `hypersimp()` simplifies more expressions if the term is `hypergeometric`.
  * Added more tests to ensure if `hypersimp()` gives correct result for all `hypergeometric expressions`.

<!-- END RELEASE NOTES -->

Harikrishna-Srinivasan committed:
Improve hypersimp, added basic tests

Harikrishna-Srinivasan committed:
Merge branch 'master' of https://github.com/Harikrishna-Srinivasan/sympy

Harikrishna-Srinivasan committed:
Corrected few tests

Harikrishna-Srinivasan committed:
Removed the typehints

oscarbenjamin commented on sympy/simplify/simplify.py:
```diff
@@ -315,17 +315,59 @@ def hypersimp(f, k):
+                return b**c
+
+    elif isinstance(f, gamma):
+        from sympy.functions.combinatorial.factorials import RisingFactorial
```
Can this import not be at the top of the module?

oscarbenjamin commented:
It looks like there is a bug in `Piecewise._eval_nseries`:
https://github.com/sympy/sympy/blob/3a4a13236c07c1a025b857777fa7e69f10824b01/sympy/functions/elementary/piecewise.py#L750-L752
It should pass the cdir parameter through.

That bug is showing up as failing tests though presumably because `hypersimp` is returning something different in some cases. It should be investigated firstly whether `hypersimp` is returning correct results in those cases. Then also the `_eval_nseries` method should be fixed.

I think that the other failing tests can just be changed although I haven't looked closely at each.

Harikrishna-Srinivasan commented:
With the new (& proper) assumption that `a**(a*k)` is hypergeometric, now we have one more test to be added to `XFAIL` (which will pass after this PR) in `test_recurr.py`.
```python
@XFAIL
def test_issue_27975c():
    assert rsolve_hyper([-2, 1], 2**(a*n + 1), n) is not None
```
From [https://github.com/sympy/sympy/issues/27975 (comment)](https://github.com/sympy/sympy/issues/27975#issuecomment-2818690333), you suggested me to use `_linear_eq_to_dict`, but I'm gonna use slightly more efficient `_lin_eq2dict` (are there any drawbacks to it?)

But `together()` in place of `factor_terms()` doesn't seem to give results properly.
With,
```diff
@@ -321,7 +323,7 @@ def hypersimp(f, k):
     f = f.rewrite(gamma)

     f = expand_func(f)
-    f = factor_terms(f)
+    f = together(f)
     return _get_term_ratio(f, k)
```
I got,
```python
In [1]: from sympy import *

In [2]: k = symbols('k')

In [3]: term = -2**(1 - k)*factorial(k + 2)/factorial(k - 1) + factorial(k + 3)/(2**k*factorial(k))
   ...: print(hypersimp(term, k))
None
```
But expected is (as per test case),
```python
In [3]: term = -2**(1 - k)*factorial(k + 2)/factorial(k - 1) + factorial(k + 3)/(2**k*factorial(k))
   ...: print(hypersimp(term, k))
(2 - k)*(k + 3)/(2*(3 - k)*(k + 1))
```
After `together(f)`, f became, `(k + 1)*(k + 2)*(-2**k*2**(1 - k)*k + k + 3)/2**k`,
but after `factor_terms(f)` it became, `(3 - k)*(k + 1)*(k + 2)/2**k`, here both are equivalent.
```python
In [7]: ((k + 1)*(k + 2)*(-2**k*2**(1 - k)*k + k + 3)/2**k).equals((3 - k)*(k + 1)*(k + 2)/2**k)
Out[7]: True
```
Yet, it will not work.
So may be `factor_terms()` is suitable here.

And, what should we do about bug in `Piecewise._eval_nseries`?

oscarbenjamin commented:
> I'm gonna use slightly more efficient `_lin_eq2dict` (are there any drawbacks to it?)

That's reasonable. There are no drawbacks.

> But `together()` in place of `factor_terms()` doesn't seem to give results properly.
> With,

I mean to use `together` on the output of `hypersimp`. Probably it is only needed for the rational function case so perhaps that case could use `together` in get_ratio like `return together(f.subs(k, k+1) / f)`.

> And, what should we do about bug in `Piecewise._eval_nseries`?

First identify what it is that has changed that causes the bug to show up. Somehow hypersimp is returning a different result in some cases. Potentially that means that `hypersimp` is returning incorrect results or could be improved. We need to see what are the inputs to hypersimp and then compare how hypersimp from master branch performs compared to hypersimp in the PR for those examples.

Secondly fix the bug in `_eval_nseries`. It just needs `cdir=cdir`.

Harikrishna-Srinivasan commented:
> I mean to use together on the output of hypersimp. Probably it is only needed for the rational function case so perhaps that case could use together in get_ratio like return together(f.subs(k, k+1) / f)

What I think you meant,
```python
def _get_term_ratio(f, k):
    if f.is_rational_function(k):
        # If f is a rational function then its term ratio is as well
        return together(f.subs(k, k+1) / f) # here ??
    ...
```
If this is what you meant, the code works just fine without this also, but this simplifies much, which is good. So I will keep this.

And, I also changed this test (I'm not sure if 3rd assertion change is acceptable)
```diff
diff --git a/sympy/stats/tests/test_discrete_rv.py b/sympy/stats/tests/test_discrete_rv.py
index 39650a1a08..c1f911972f 100644
--- a/sympy/stats/tests/test_discrete_rv.py
+++ b/sympy/stats/tests/test_discrete_rv.py
@@ -305,8 +305,12 @@ def test_product_spaces():
     X1 = Geometric('X1', S.Half)
     X2 = Geometric('X2', Rational(1, 3))
     assert str(P(X1 + X2 < 3).rewrite(Sum)) == (
-        "Sum(Piecewise((1/(4*2**n), n >= -1), (0, True)), (n, -oo, -1))/3")
+        'Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, X2 - n <= 2), '
+        '(0, True)), (X2, 1, oo), (n, -oo, -1))')
     assert str(P(X1 + X2 > 3).rewrite(Sum)) == (
-        'Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, '
-        'X2 - n <= 2), (0, True)), (X2, 1, oo), (n, 1, oo))')
-    assert P(Eq(X1 + X2, 3)) == Rational(1, 12)
+        'Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, X2 - n <= 2), (0, True)), '
+        '(X2, 1, oo), (n, 1, oo))')
+    assert str(P(Eq(X1 + X2, 3))) == (
+        'Sum(Piecewise((2**(X2 - 2)*(2/3)**(X2 - 1)/6, X2 <= 2), (0, True)), '
+        '(X2, 1, oo))'
+    )
```
In `core/expr.py` shall I add `def _eval_nseries(self, x, n, logx, cdir=0)`, cdir=0 as default. Still, `cdir=cdir` is needed in `_eval_nseries` in `functions/elementary/piecewise.py` (as you said).

Harikrishna-Srinivasan committed:
linear_coeff improved, tests updated, typerror fixed _eval_nseries

oscarbenjamin commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -748,7 +748,7 @@ def _solve_relational(r):
         return True, list(uniq(int_expr))
 
     def _eval_nseries(self, x, n, logx, cdir=0):
-        args = [(ec.expr._eval_nseries(x, n, logx), ec.cond) for ec in self.args]
+        args = [(ec.expr._eval_nseries(x, n, logx, cdir=cdir), ec.cond) for ec in self.args]
         return self.func(*args)
```
What are the examples involving hypersimp that caused this method to get called?

Harikrishna-Srinivasan replied:
Actually `hypersimp` does not cal it at all. But in `test_discrete_rv.py` test case,
```python
================================== FAILURES ===================================
_______________________________ test_DiscreteRV _______________________________

    def test_DiscreteRV():
        p = S(1)/2
        x = Symbol('x', integer=True, positive=True)
        pdf = p*(1 - p)**(x - 1) # pdf of Geometric Distribution
        D = DiscreteRV(x, pdf, set=S.Naturals, check=True)
>       assert E(D) == E(Geometric('G', S(1)/2)) == 2

sympy\stats\tests\test_discrete_rv.py:206: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sympy\stats\rv.py:808: in expectation
    return Expectation(expr, condition).doit(**kwargs)
sympy\stats\symbolic_probability.py:288: in doit
    result = pspace(expr).compute_expectation(expr, evaluate=evaluate)
sympy\stats\drv.py:313: in compute_expectation
    return self.distribution.expectation(expr, x, evaluate=evaluate,
sympy\stats\drv.py:162: in expectation
    taylor = poly(series(mgf, t, 0, deg + 1).removeO(), t)
sympy\series\series.py:63: in series
    return expr.series(x, x0, n, dir)
sympy\core\expr.py:3084: in series
    rv = self.subs(x, xpos).series(xpos, x0, n, dir, logx=logx, cdir=cdir)
sympy\core\expr.py:3092: in series
    s1 = self._eval_nseries(x, n=n, logx=logx, cdir=cdir)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Piecewise((exp(_x)/(2*(1 - exp(_x)/2)), exp(_x) < 2), (Sum(exp(_x*_x)/2**_x, (_x, 1, oo)), True))
x = _x, n = 2, logx = None, cdir = 1

    def _eval_nseries(self, x, n, logx, cdir=0):
>       args = [(ec.expr._eval_nseries(x, n, logx), ec.cond) for ec in self.args]
E       TypeError: Expr._eval_nseries() missing 1 required positional argument: 'cdir'

sympy\functions\elementary\piecewise.py:751: TypeError
```
This is corrected.

oscarbenjamin replied:
Several test cases failed. Were they all because of hypersimp simplifying this same example?

oscarbenjamin replied:
What is hypersimp called with in the case of test_discreterv?

oscarbenjamin commented on sympy/solvers/tests/test_recurr.py:
```diff
@@ -282,6 +281,10 @@ def test_issue_27975b():
 
+@XFAIL
+def test_issue_27975c():
+    assert rsolve_hyper([-2, 1], 2**(a*n + 1), n) is not None
```
I think this one returns correct output so it should not be an XFAIL test.

Harikrishna-Srinivasan replied:
All three a, b & c passes the case, so shall I change it from `is not None` to the expectd result?

oscarbenjamin replied:
If the function returns a correct result then the test should be to check for that correct result.

oscarbenjamin commented:
In the case of test_poisson_CompoundDist what is the changed hypersimp output? Is it similar to test_discreterv?

It is perhaps worth having a fallback case for `get_term_ratio` so that instead of returning None it checks this:
```python
g = f.subs(k, k+1) / f
if g.is_rational_function(k):
    return together(g)
```

Harikrishna-Srinivasan commented:
> In the case of test_poisson_CompoundDist what is the changed hypersimp output? Is it similar to test_discreterv?
> 
> It is perhaps worth having a fallback case for `get_term_ratio` so that instead of returning None it checks this:

That's okay, but it passes `test_discrete_rv` (test_poisson_CompoundDist) already without this also.

oscarbenjamin commented:
> > In the case of test_poisson_CompoundDist what is the changed hypersimp output? Is it similar to test_discreterv?
> 
> I observed no change at all.

Then why did the test fail before?

https://github.com/sympy/sympy/actions/runs/14623117400/job/41028094337

> > It is perhaps worth having a fallback case for `get_term_ratio` so that instead of returning None it checks this:
> 
> That's okay, but it passes `test_discrete_rv` (test_poisson_CompoundDist) already without this also.

The `test_product_space` test that failed is because of not checking `f.subs(k, k+1)`.

Harikrishna-Srinivasan commented:
> Then why did the test fail before?

I am not sure why, but here is the screenshot of the test that I ran in my laptop with the previous code.

<img width="911" alt="image" src="https://github.com/user-attachments/assets/26f26cf7-9b68-42ef-ab0e-26bffcb1cf41" />

This has happened to me in previous PR in `https://github.com/sympy/sympy/pull/27931/commits/b1e52a90fd4435cad7cb2f9589d1a5f131652925` I think; no test case should have failed but some failed.
I made some other changes & pushed again, it worked.

Harikrishna-Srinivasan committed:
Added passed tests

oscarbenjamin commented:
> > Then why did the test fail before?
> 
> I am not sure why

It still fails now if you remove the cdir=cdir from piecewise.py:
```
self = Piecewise((gamma(k)/(-t*exp(_x)/(t + 1) + 1)**k, ((k >= 1) & (t*exp(_x)/(t + 1) < 1)) | ((k < 1) & (t*exp(_x)/(t + 1) <= 1) & Ne(t*exp(_x)/(t + 1), 1))), (Sum(t**_x*exp(_x*_x)*gamma(_x + k)/((t + 1)**_x*factorial(_x)), (_x, 0, oo)), True))
x = _x, n = 2, logx = None, cdir = 1

    def _eval_nseries(self, x, n, logx, cdir=0):
>       args = [(ec.expr._eval_nseries(x, n, logx), ec.cond) for ec in self.args]
E       TypeError: Expr._eval_nseries() missing 1 required positional argument: 'cdir'

sympy/functions/elementary/piecewise.py:751: TypeError
                                          DO *NOT* COMMIT!                                          
====================================== short test summary info ======================================
FAILED sympy/stats/tests/test_compound_rv.py::test_poisson_CompoundDist - TypeError: Expr._eval_nseries() missing 1 required positional argument: 'cdir'
=================================== 1 failed, 5 passed in 24.58s ====================================
```

> I made some other changes & pushed again, it worked.

That is because you added cdir=cdir. Let me remind you what I said before:

> > And, what should we do about bug in `Piecewise._eval_nseries`?
> 
> First identify what it is that has changed that causes the bug to show up. Somehow hypersimp is returning a different result in some cases. Potentially that means that `hypersimp` is returning incorrect results or could be improved. We need to see what are the inputs to hypersimp and then compare how hypersimp from master branch performs compared to hypersimp in the PR for those examples.
> 
> Secondly fix the bug in `_eval_nseries`. It just needs `cdir=cdir`.

You skipped the first part and went to the second. That is why I am now asking all these questions.

Harikrishna-Srinivasan committed:
Corrected tests for rsolve_hyper

oscarbenjamin commented:
> In the case of test_poisson_CompoundDist what is the changed hypersimp output? Is it similar to test_discreterv?

The answer is this:
```python
In [1]: hypersimp(t**k*exp(k*z)*gamma(2*k)/(k*(t + 1)**k*gamma(k)), k)
Out[1]: 
                 z
2⋅k⋅t⋅(2⋅k + 1)⋅ℯ 
──────────────────
 (k + 1)⋅(t + 1)  

In [2]: hypersimp(exp(k*z), k)
Out[2]: 
 z
ℯ 
```
On master both of these give None and it works only if the exponential factor is removed:
```python
In [1]: hypersimp(t**k*exp(k*z)*gamma(2*k)/(k*(t + 1)**k*gamma(k)), k)

In [2]: hypersimp(exp(k*z), k)

In [3]: hypersimp(t**k*gamma(2*k)/(k*(t + 1)**k*gamma(k)), k)
Out[3]: 
2⋅k⋅t⋅(2⋅k + 1)
───────────────
(k + 1)⋅(t + 1)
```
The new output is correct and is better than None.

Harikrishna-Srinivasan commented:
Oh! okay, thanks.

Harikrishna-Srinivasan committed:
Removed unused imports

oscarbenjamin commented on sympy/stats/tests/test_discrete_rv.py:
```diff
@@ -305,8 +305,12 @@ def test_product_spaces():
     X1 = Geometric('X1', S.Half)
     X2 = Geometric('X2', Rational(1, 3))
     assert str(P(X1 + X2 < 3).rewrite(Sum)) == (
-        "Sum(Piecewise((1/(4*2**n), n >= -1), (0, True)), (n, -oo, -1))/3")
+        'Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, X2 - n <= 2), '
+        '(0, True)), (X2, 1, oo), (n, -oo, -1))')
```
This test would not need to be changed if `(f.subs(k, k+1) / f).is_rational_function(k)` was checked as a fallback in get_term_ratio.

Harikrishna-Srinivasan replied:
Even though you didn't mean this it works (and passes the old test):
```python
def hypersimp(f, k):
    ...
    f = expand_func(f)
    f = factor_terms(f)
    g = f.subs(k, k + 1) / f

    if g.is_rational_function(k): # shall I add it here?
        return together(g)
    return _get_term_ratio(f, k)
```
instead of 
```python
def _get_term_ratio(f, k):
    if f.is_rational_function(k):
        # If f is a rational function then its term ratio is as well
        return together(f.subs(k, k+1) / f)

    g = f.subs(k, k + 1) / f
    if g.is_rational_function(k):
        return together(g)
    ...
```

oscarbenjamin replied:
That won't work when you have a product of a rational function and something else. Also we only want to compute `f.subs(k, k+1)` only if necessary i.e. after checking everything else.

Harikrishna-Srinivasan commented:
I got a failed test case
```python
rf(n, k).is_hypergeometric(k)
```
as `False`
Because of this:
```
f: Piecewise((gamma(k + n)/gamma(n), n > 0), ((-1)**k*n*gamma(-n)/((k + n)*gamma(-(k + n))), True))
Expected: k + n
Got: None
```
It called `hypersimp(f, k)`,and got `None` as we are not handling `Piecewise` cases appropriately.

Harikrishna-Srinivasan committed:
Add ratio is rational as fallback

oscarbenjamin commented:
It could be possible to handle Piecewise here but probably in the first instance `rf` should be handled explicitly rather than by rewriting to Piecewise. The rewrite should exclude rf and then get_term_ratio can handle it directly.

More generally to handle Piecewise though the way to do it is like:
```python
In [21]: p = Piecewise((gamma(k + n)/gamma(n), n > 0), ((-1)**k*n*gamma(-n)/((k + n)*gamma(-(k + n))), True))

In [22]: p
Out[22]:
⎧    Γ(k + n)
⎪    ────────       for n > 0
⎪      Γ(n)
⎪
⎨      k
⎪  (-1) ⋅n⋅Γ(-n)
⎪─────────────────  otherwise
⎪(k + n)⋅Γ(-k - n)
⎩

In [23]: p.args
Out[23]:
⎛                   ⎛      k                ⎞⎞
⎜⎛Γ(k + n)       ⎞  ⎜  (-1) ⋅n⋅Γ(-n)        ⎟⎟
⎜⎜────────, n > 0⎟, ⎜─────────────────, True⎟⎟
⎝⎝  Γ(n)         ⎠  ⎝(k + n)⋅Γ(-k - n)      ⎠⎠

In [24]: any(c.has_xfree({k}) for e, c in p.args)
Out[24]: False

In [25]: hypersimp(p.args[0][0], k)
Out[25]: k + n

In [26]: hypersimp(p.args[1][0], k)
Out[26]:
-(k + n)⋅(-(k + n) - 1)
────────────────────────
       k + n + 1

In [27]: simplify(_)
Out[27]: k + n
```
If the conditions don't depend on `k` then it is acceptable to computed the term ratio of each case and return the result as a Piecewise. Here all cases would be the same and the Piecewise simplifies away:
```python
In [34]: Piecewise(*[(hypersimp(e, k), c) for e, c in p.args])
Out[34]:
⎧         k + n            for n > 0
⎪
⎨-(k + n)⋅(-(k + n) - 1)
⎪────────────────────────  otherwise
⎩       k + n + 1

In [35]: Piecewise(*[(hypersimp(e, k).simplify(), c) for e, c in p.args])
Out[35]: k + n
```
It would be worth handling Piecewise but that is not the right way to deal with `rf` I think so in the first instance `rf` should be handled directly.

Harikrishna-Srinivasan commented:
But wouldn't this be even better:
```python
def _linear_coeff(e, k):
    try:
        return _lin_eq2dict(e, {k})[1].get(k, S.Zero)
    except (PolyNonlinearError, ValueError):
        return None

def is_rational_ratio(expr, k):
    fns = {gamma, exp, Pow, Piecewise}
    if expr.is_rational_function(k):
        return True
    args = expr.args
    if expr.func not in fns:
        return all(is_rational_ratio(arg, k) for arg in args)
    if expr.func == Piecewise:
        return all(is_hyper(c, k) for c, _ in args)
    if expr.func == gamma:
        return _linear_coeff(args[0], k) is not None
    if expr.func in {Pow, exp}:
        base, expn = args
        if base.has_xfree({k}):
            return expn.is_rational
        coeff = _linear_coeff(expn, k)
        return coeff is not None and coeff.is_rational
    return False

def is_hyper(f, k):
    f = f.rewrite(gamma)
    for arg in Add.make_args(f):
        numer, denom = fraction(arg)
        numer = numer.args if isinstance(numer, Mul) else (numer,)
        for term in numer:
            if not is_rational_ratio(term, k):
                return False
        denom = denom.args if isinstance(denom, Mul) else (denom,)
        for term in denom:
            if not is_rational_ratio(term, k):
                return False
    return True
```

And, the problem `Piecewise((gamma(k + n)/gamma(n), n > 0), ((-1)**k*n*gamma(-n)/((k + n)*gamma(-(k + n))), True))` came only becase of `.rewrite(gamma)` which is unavoidable.

So, the only way I can think is:
```python

def _get_term_ratio(f, k):
    if f.is_rational_function(k):
        # If f is a rational function then its term ratio is as well
        return together(f.subs(k, k+1) / f)

    elif isinstance(f, Piecewise):
        # Compute term ratio for each piece
        if any(interval.has_xfree({k}) for (v, interval) in f.args):
            return None

        ratios = set()
        for a in f.args:
            r = _get_term_ratio(a[0], k)
            if r is None:
                return None
            ratios.add((r, a[1]))
        return Piecewise(*ratios)
    ...
```

oscarbenjamin commented:
> the problem `Piecewise((gamma(k + n)/gamma(n), n > 0), ((-1)**k*n*gamma(-n)/((k + n)*gamma(-(k + n))), True))` came only becase of `.rewrite(gamma)` which is unavoidable.

It is not unavoidable. It is possible with `rewrite` to specify which kinds of expressions should be rewritten so if we don't want to rewrite `rf` we don't need to.

oscarbenjamin commented:
> But wouldn't this be even better:

Are you suggesting to change the `is_hypergeometric` function or the `hypersimp` function?

It is necessary to check the conditions in the Piecewise to see that they are free of `k` e.g.:
```python
In [1]: p = Piecewise((0, k<10), (1, True))

In [2]: p
Out[2]: 
⎧0  for k < 10
⎨             
⎩1  otherwise 

In [3]: hypersimp(p, k) # incorrect
Out[3]: 1

In [4]: p2 = Piecewise((0, k>10), (1, True))

In [6]: p2
Out[6]: 
⎧0  for k > 10
⎨             
⎩1  otherwise 

In [7]: hypersimp(p2, k) # also incorrect
Out[7]: 1
```

Harikrishna-Srinivasan commented:
I'm talking about `is_hypergeomtric` only which is in `basic.py`. We can change this. 

> It is necessary to check the conditions in the Piecewise to see that they are free of k e.g.:

Got it.

oscarbenjamin commented:
> I'm talking about `is_hypergeomtric` only which is in `basic.py`. We can change this.

I think that the only change there should be to remove the `Piecewise` check. If `hypersimp` returns correct results then `is_hypergeometric` should be able to just call it.

Harikrishna-Srinivasan commented:
I searched thoroughly in the docs, I didn't find how I can `rewrite`, excluding some `func`.

So, should I do something like this or should I create my own `_eval_rewrite_as_gamma`?
```python
f = f.rewrite([
    factorial,
    factorial2,
    binomial,
    beta,
], gamma)
```

I cannot check only for `isisntance(f, rf)` & then leave it without rewriting the other terms. Then I'll be rewriting `rf(n, k) + factorial(k)` which will result in same issue. If I check for `f.has(rf)`, then I'll end up leaving the `factorial(k)` term untouched, which is also incorrect.

Moreover, I'm chnging to this:
```python
def _get_term_ratio(f, k):
    if f.is_rational_function(k):
        # If f is a rational function then its term ratio is as well
        return together(f.subs(k, k+1) / f)

    elif isinstance(f, Mul):
        # Compute term ratio for each factor
        ratios = []
        for a in f.args:
            r = _get_term_ratio(a, k)
            if r is None:
                return None
            ratios.append(r)
        return Mul(*ratios)

    elif isinstance(f, (Pow, exp)):
        b, e = f.as_base_exp()
        if e.is_Integer:
            # f(k)**e, Compute term ratio for f(k)
            r = _get_term_ratio(b, k)
            if r is not None:
                return r**e
        elif not b.has_xfree({k}):
            # b**(c*k + ...), ratio is b**c
            c = _linear_coeff(e, k)
            if c is not None:
                return b**c

    elif isinstance(f, gamma):
        a = f.args[0]
        c = _linear_coeff(a, k)
        if c is not None and c.is_Integer:
            # gamma(a) = gamma(c*k + ...)
            return RisingFactorial(a, c)

    else:
        g = f.subs(k, k+1)/f
        if g.is_rational_function(k):
            return together(g)

        elif isinstance(f, Piecewise):
            # Compute term ratio for each piece
            f = piecewise_simplify(f)
            if any(interval.has_xfree({k}) for _, interval in f.args):
                return None

            ratios = []
            for a in f.args:
                r = _get_term_ratio(a[0], k)
                if r is None:
                    return None
                ratios.append((r, a[1]))
            res = Piecewise(*ratios)
            if res.is_rational_function(k):
                return res

    return None
```
seems to pass all tests, I'll check thoroughly and add more tests.

oscarbenjamin commented:
> I didn't find how I can `rewrite`, excluding some `func`.

I don't think there is a way but as you say you can just list the things that should be rewritten. I'm not sure that anything should be rewritten at the outset though. It might be better to have something that explicitly handles all cases like `factorial` etc and only starts using rewrite, factor_terms etc when that fails.

It would be worth calling `piecewise_fold` before checking `isinstance(..., Piecewise)`.

Harikrishna-Srinivasan commented:
```python
In [1]: from sympy import *

In [2]: f = (Rational(2, 3) * (Rational(2, 3))**Dummy('d') * Piecewise(
   ...:     (Rational(1, 4) * 2**(-Dummy('d')) * 2**Dummy('d'), Dummy('d') - Dummy('d') <= 0),
   ...:     (0, True)
   ...: ))
   ...: hypersimp(f, Dummy('d'))

f:  2*(2/3)**_d*Piecewise((2**_d/(4*2**_d), _d - _d <= 0), (0, True))/3 
Expected:  None 
Got:  Piecewise((2**_d/(4*2**_d), _d - _d <= 0), (0, True))*Piecewise((4*2**_d/2**_d, _d - _d <= 0), (zoo, True))

Out[2]:
⎛⎧ -d  d               ⎞
⎜⎪2  ⋅2                ⎟ ⎛⎧   -d  d               ⎞
⎜⎪──────  for d - d ≤ 0⎟ ⎜⎪4⋅2  ⋅2   for d - d ≤ 0⎟
⎜⎨  4                  ⎟⋅⎜⎨                       ⎟
⎜⎪                     ⎟ ⎜⎪  zoo       otherwise  ⎟
⎜⎪  0       otherwise  ⎟ ⎝⎩                       ⎠
⎝⎩                     ⎠
```

Atleast, `_d - _d` must simplify to True, so is this output correct?
I checked & found that `f.is_rational_function(_d)` is also True.

(I know that ``Dummy('d') is not Dummy('d')``)

But `(f.subs(_d, _d+1) / f).simplify(ratio=S.Infinity)` returns,
```python
In[5]: (f.subs(_d, _d+1) / f).simplify(ratio=S.Infinity)`In [5]: (f.subs(_d, _d+1) / f).simplify(ratio=S.Infinity)
Out[5]: 
⎧ 1   for d ≤ d
⎨
⎩nan  otherwise
```
Which of these should I actually return (that is correct).

The code I used was,
```python
def _get_term_ratio(f, k):
    if f.is_rational_function(k):
        # If f is a rational function then its term ratio is as well
        return together(f.subs(k, k+1) / f)

    elif isinstance(f, Mul):
        # Compute term ratio for each factor
        ratios = []
        for a in f.args:
            r = _get_term_ratio(a, k)
            if r is None:
                return None
            ratios.append(r)
        return Mul(*ratios)
    
    elif isinstance(f, RisingFactorial):
        return f.args[1] + f.args[0]
    
    elif isinstance(f, FallingFactorial):
        return f.args[1] - f.args[0]

    elif isinstance(f, (Pow, exp)):
        b, e = f.as_base_exp()
        if e.is_Integer:
            # f(k)**e, Compute term ratio for f(k)
            r = _get_term_ratio(b, k)
            if r is not None:
                return r**e
        elif not b.has_xfree({k}):
            # b**(c*k + ...), ratio is b**c
            c = _linear_coeff(e, k)
            if c is not None:
                return b**c

    elif isinstance(f, gamma):
        a = f.args[0]
        c = _linear_coeff(a, k)
        if c is not None and c.is_Integer:
            # gamma(a) = gamma(c*k + ...)
            return RisingFactorial(a, c)

    else:
        g = f.subs(k, k+1)/f
        if g.is_rational_function(k):
            return together(g)

        elif isinstance(f, Piecewise):
            # Compute term ratio for each piece
            f = piecewise_fold(f)
            if any(interval.has_xfree({k}) for _, interval in f.args):
                return None

            ratios = set()
            for a in f.args:
                r = _get_term_ratio(a[0], k)
                if r is None:
                    return None
                ratios.add((r, a[1]))
            res = Piecewise(*ratios)
            if res.is_rational_function(k):
                return res

    return None
```
And, `f = f.rewrite([
            factorial,
            factorial2,
            binomial,
            beta
        ], gamma)` was the only change in `hypersimp(f, k)`.

oscarbenjamin commented:
> Atleast, `_d - _d` must simplify to True, so is this output correct?

The expression has two different dummy symbols in it which is confusing.

The expression that you created by calling Dummy many times has more than two dummy symbols and is not equivalent.

Harikrishna-Srinivasan commented:
Yeah, I got confused when I saw `d - d = 0`  (I though both were the same).

Anyways, but I don't understand, whether we should `simplify` at last to return like, 

```python
⎧ 1   for d ≤ d
⎨
⎩nan  otherwise
```
Or, can we leave it as such?

oscarbenjamin commented:
> Anyways, but I don't understand, whether we should `simplify` at last to return like,

You are going to have to explain more clearly what this example is. Figure out how to make the expression that is being passed to hypersimp without using dummy symbols and then demonstrate the code to make it here and clearly explain what your question is.

oscarbenjamin commented:
I'm not sure about that case. It is a rational function of `g` but we don't know if it is actually identically equal to zero. I think it is better if hypersimp just returns None for this example but I am not sure what exactly is the right way to express that hypersimp should just reject it. Maybe just that it has a Piecewise that was not simplified away by calling hypersimp on its expressions.

Anyway I don't think that is what is happening in the test case you showed before with `_d` because I think that has two dummy symbols not 6. There the symbol does appear in the Piecewise.

Harikrishna-Srinivasan committed:
Corrected hypersimp for rf, ff

Harikrishna-Srinivasan committed:
Correctly added factor_terms, expand_func

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master

Harikrishna-Srinivasan commented:
> I think it is better if hypersimp just returns None

Why do you think so?
Isn't this still `hypergeometric`? (It is)

Will returning `None` if `f` has `Piecewise` in the input itself. Will this be correct, probably not.
But I'm wondering why we have,
```python
    def is_hypergeometric(self, k):
        from sympy.simplify.simplify import hypersimp
        from sympy.functions.elementary.piecewise import Piecewise
        if self.has(Piecewise): # not even checks if self is rational function of k?????
            return None
        return hypersimp(self, k) is not None
```
And, master returns `None` (I think) for most of Piecewise inputs.
This is because of this, (In `Expr` class of expr.py)
```python
    def _eval_is_rational_function(self, syms) -> bool | None:
        if self in syms:
            return True
        if not self.has_xfree(syms):
            return True
        # subclasses should return True or False
        # Piecewise didn't overrride it. The issue is here.
        return None
```

In short, `is_rational_function(k)` calls `self._eval_is_rational_function(syms)` which returned `None`.

**The issue is that `Piecewise` did not override this function, so I'm getting None for,**

```python
f = Piecewise(
    (g, e <= 0),
    (-g, e > 0)
)
hypersimp(f, g)
```
returned `None`. (in my branch). (I check if `res.is_rational_function(k)` inside _get_rational_ratio)

And, I think `simplify()` may be necessary, (in Piecewise case only)
```python
In [2]: f = Piecewise(
   ...:     (Rational(1, 4), e <= 0),
   ...:     (1, e > 0)
   ...: )
   ...: hypersimp(f, g)
Out[2]: 
⎛⎧1/4  for e ≤ 0⎞ ⎛⎧4  for e ≤ 0⎞
⎜⎨              ⎟⋅⎜⎨            ⎟
⎝⎩ 1   otherwise⎠ ⎝⎩1  otherwise⎠
```
But this seems better, (master returns `None` only)
```python
In [3]: hypersimp(f, g).simplify()
Out[3]: 1
```

Harikrishna-Srinivasan committed:
Added a few tests, improved Piecewise handling

Harikrishna-Srinivasan commented on sympy/simplify/simplify.py:
```diff
@@ -315,18 +319,91 @@ def hypersimp(f, k):
-        g = g.args[-1][0]
-    g = expand_func(g)
-    g = powsimp(g, deep=True, combine='exp')
+    f = f.rewrite([
```
The previous test on `hypersimp(rf(n, k), k)` failed because I did not handle `piecewise` properly.

Now it works with `f = f.rewrite(gamma)` I think it would be better if I remove this change & leave it as `f = f.rewrite(gamma)`.
We may have left a few more special cases & may not have handled them appropriately.

Harikrishna-Srinivasan committed:
Correctd piecewise handling @ piecewise_exclusive

Harikrishna-Srinivasan committed:
Added necessary imports

Harikrishna-Srinivasan committed:
corrected test_simplify.py tests

Harikrishna-Srinivasan committed:
removed unused import

Harikrishna-Srinivasan committed:
removed unused import

Harikrishna-Srinivasan committed:
Corrected piecewise interval check

Harikrishna-Srinivasan commented:
> If the conditions don't depend on k then it is acceptable to computed the term ratio of each case and return the result as a Piecewise

As you said in https://github.com/sympy/sympy/pull/27984#issuecomment-2830627329, I think
```python
In [24]: any(c.has_xfree({k}) for e, c in p.args)
```
is only necessary after computing the `piecewise simplifications`.

`hypersimp(Piecewise((gamma(n + 1)/gamma(-k + n + 1), n >= 0), ((-1)**k*gamma(k - n)/gamma(-n), True)), n)` returns None, but expected is `-n/(k - n + 1)`.

So I think if piecewise intervals depend on `k` doesn't directly mean it's not hypergeometric, but this should be checked after some simplifications.
```diff
diff --git a/sympy/simplify/simplify.py b/sympy/simplify/simplify.py
index 65e3c13a93..7ee44ca03a 100644
--- a/sympy/simplify/simplify.py
+++ b/sympy/simplify/simplify.py
@@ -374,9 +374,6 @@
def _get_term_ratio(f, k):
             if not isinstance(f, Piecewise):
                 return _get_term_ratio(f, k)

-            if any(interval.has_xfree({k}) for _, interval in f.args):
-                return None
-
             ratios = set()
             for a in f.args:
                 r = _get_term_ratio(a[0], k)
@@ -384,6 +381,9 @@
def _get_term_ratio(f, k):
                     return None
                 ratios.add((r, a[1]))
             res = Piecewise(*ratios)
+            if isinstance(res, Piecewise) and any(interval.has_xfree({k}) for _, interval in res.args):
+                return None
+
             return res.simplify()

     return None
```
You also suggested me to implement `ff` & `rf` cases separately. I don't think it was a good idea to do so, because rewriting in `gamma` excluding these might miss some special functions also (due to human error).

So I removed that and implemented `piecewise` case correctly.

Harikrishna-Srinivasan committed:
Simplify before interval check

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master

Harikrishna-Srinivasan committed:
Merge branches 'master' and 'master' of https://github.com/Harikrishna-Srinivasan/sympy

Harikrishna-Srinivasan committed:
corrected piecewise test

oscarbenjamin commented:
> You also suggested me to implement `ff` & `rf` cases separately. I don't think it was a good idea to do so, because rewriting in `gamma` excluding these might miss some special functions also (due to human error).

I'm not sure what you mean by this.

Harikrishna-Srinivasan committed:
Added .equals for lower python versions

Harikrishna-Srinivasan commented:
> I'm not sure what you mean by this.

Okay let me explain briefly.
You suggested me to separate `ff` & `rf` like

> It could be possible to handle Piecewise here but probably in the first instance rf should be handled explicitly rather than by rewriting to Piecewise. The rewrite should exclude rf and then get_term_ratio can handle it directly.

in https://github.com/sympy/sympy/pull/27984#issuecomment-2830627329
```python
def _get_term_ratio(f, k):
  ...
  elif isinstance(f, RisingFactorial):
        return f.args[1] + f.args[0]
    
    elif isinstance(f, FallingFactorial):
        return f.args[1] - f.args[0]
...
```

And also
```python
f = f.rewrite([
        factorial,
        factorial2,
        binomial,
        beta ], gamma) # instead of f = f.rewrite(gamma)
```
I changed both as they are more prone to error. I mean I might miss some cases.

Harikrishna-Srinivasan commented on sympy/simplify/tests/test_simplify.py:
```diff
@@ -360,14 +361,60 @@ def test_hypersimp():
+    assert hypersimp(term, c) == Piecewise((4*2**b/2**a, a - b <= 0), (zoo, True)) * \
+        Piecewise((2**a/(4*2**b), a - b <= 0), (0, True))
+    assert hypersimp(Piecewise((k, True)), k) == (k + 1) / k
+    assert hypersimp(Piecewise((k + 1, n < 2), (i*k, True)), k).equals(Piecewise(((k + 1)/k, n >= 2), ((k + 2)/(k + 1), True)))
```
But I'm not sure why this test previously failed in python 3.9, 3.11 but passed in 3.12, ...? (When `==` was used)
So I changed to `.equals()` intead of `==`.

oscarbenjamin replied:
We should find out why that happened. Can you change it back to `==`?

Harikrishna-Srinivasan committed:
Changed to ==

oscarbenjamin commented on sympy/simplify/simplify.py:
```diff
@@ -315,18 +317,83 @@ def hypersimp(f, k):
+            if not isinstance(f, Piecewise):
+                return _get_term_ratio(f, k)
+
+            ratios = set()
+            for a in f.args:
+                r = _get_term_ratio(a[0], k)
+                if r is None:
+                    return None
+                ratios.add((r, a[1]))
+            res = Piecewise(*ratios).simplify()
```
You can't use a set here because the args to a Piecewise need to be ordered.

Harikrishna-Srinivasan replied:
Sorry, I forgot that I used set here (I initially thought the order wouldn't have mattered)

Harikrishna-Srinivasan committed:
Piecewise case use list rather than set

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master

oscarbenjamin commented on sympy/simplify/simplify.py:
```diff
@@ -315,18 +317,83 @@ def hypersimp(f, k):
+
+        elif isinstance(f, Piecewise):
+            # Compute term ratio for each piece
+            f = piecewise_exclusive(f, skip_nan=True)
```
I don't think that we want to use `piecewise_exclusive` here. Why is it being used?

Harikrishna-Srinivasan replied:
You're right, using `piecewise_exclusive` will handle the interval by explicitly writing the `True` (otherwise) case appropriately.

(A sample from the output of `hypersimp` that is not shown/needed here)
```python
In [2]: piecewise_exclusive(Piecewise(((x + 1)/x, y < 2), ((x + 3)/(x + 2), y > 2), (Piecewise(((x + 1)/x, z < 2), ((x + 2)/(x + 1), True)), Eq(y, 2))))
Out[2]: 
⎧      x + 1
⎪      ─────        for y < 2
⎪        x
⎪
⎪      x + 3
⎪      ─────        for y > 2
⎪      x + 2
⎨
⎪⎧x + 1
⎪⎪─────  for z < 2
⎪⎪  x
⎪⎨                  for y = 2
⎪⎪x + 2
⎪⎪─────  for z ≥ 2
⎩⎩x + 1
```
The above one looks better than using `piecewise_fold`,
```python
In [3]: piecewise_fold(Piecewise(((x + 1)/x, y < 2), ((x + 3)/(x + 2), y > 2), (Piecewise(((x + 1)/x, z < 2), ((x + 2)/(x + 1), True)), Eq(y, 2))))
Out[3]: 
⎧x + 1
⎪─────      for y < 2
⎪  x
⎪
⎪x + 3
⎪─────      for y > 2
⎪x + 2
⎨
⎪x + 1
⎪─────  for y = 2 ∧ z < 2
⎪  x
⎪
⎪x + 2
⎪─────      for y = 2
⎩x + 1
```
Since both passes the tests, I can change it to `piecewise_fold`.
I don't have any case that will fail either.

Harikrishna-Srinivasan committed:
piecewise_exclusive to piecewise_fold

Harikrishna-Srinivasan committed:
Merge branch 'master' of https://github.com/Harikrishna-Srinivasan/sympy

oscarbenjamin commented on sympy/simplify/simplify.py:
```diff
@@ -315,18 +317,83 @@ def hypersimp(f, k):
+        if g.is_rational_function(k):
+            return together(g)
+
+        elif isinstance(f, Piecewise):
+            # Compute term ratio for each piece
+            f = piecewise_fold(f)
+            if not isinstance(f, Piecewise):
+                return _get_term_ratio(f, k)
```
The purpose of `piecewise_fold` is that it turns any expression that contains Piecewise into a Piecewise although the Piecewise might simplify away. It doesn't make sense to check if the expression is already a Piecewise before calling `piecewise_fold`:
```python
In [1]: e = 1 + Piecewise((1, x>1), (0, True))

In [2]: e
Out[2]: 
⎛⎧1  for x > 1⎞    
⎜⎨            ⎟ + 1
⎝⎩0  otherwise⎠    

In [3]: isinstance(e, Piecewise)
Out[3]: False

In [4]: e2 = piecewise_fold(e)

In [5]: e2
Out[5]: 
⎧2  for x > 1
⎨            
⎩1  otherwise

In [6]: isinstance(e2, Piecewise)
Out[6]: True
```
I'm sure if you think about this a bit you can come up with a test case that shows when `piecewise_fold` is needed.

Harikrishna-Srinivasan replied:
Actually that check is unnecessary, I confused `piecewise_fold` for `simplify()`.

In fact, even `piecewise_fold` is not needed.
It is only needed here:
```python
res = Piecewise(*ratios).simplify()
if isinstance(res, Piecewise) and any(interval.has_xfree({k}) for _, interval in res.args):
    return None
...
```
Here `.simplify()` is needed because
```python
In [1]: from sympy import *

In [2]: n, k = symbols('n k')

In [3]: p = Piecewise((k + n, n > 0), (-(k + n)*(-(k + n) - 1)/(k + n + 1), True))

In [4]: p
Out[4]: 
⎧        k + n          for n > 0
⎪
⎨(-k - n)⋅(-k - n - 1)
⎪─────────────────────  otherwise
⎩      k + n + 1

In [5]: p.simplify() # here answer may not be Piecewise
Out[5]: k + n
```

Harikrishna-Srinivasan committed:
Removed unused import

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master

oscarbenjamin commented on sympy/simplify/simplify.py:
```diff
@@ -315,18 +317,79 @@ def hypersimp(f, k):
+        if g.is_rational_function(k):
+            return together(g)
+
+        elif isinstance(f, Piecewise):
```
Testing `isinstance(f, Piecewise)` misses cases where the expression contains a Piecewise:
```python
In [11]: p = 1 + Piecewise((2**k-1, x > 0), (2**(k-1)-1, True))

In [12]: p
Out[12]:
⎛⎧   k                 ⎞
⎜⎪  2  - 1    for x > 0⎟
⎜⎨                     ⎟ + 1
⎜⎪ k - 1               ⎟
⎝⎩2      - 1  otherwise⎠

In [13]: hypersimp(p, k)

In [14]: hypersimp(piecewise_fold(p), k)
Out[14]: 2
```

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master

Harikrishna-Srinivasan committed:
Merge branch 'sympy:master' into master
